### PR TITLE
GPU Health API improvements

### PR DESCRIPTION
**Pull request by nvvfedorov**
The go-dcgm GPU health API modified to expose API access to the following DCGM API methods:

* HealthSet - Enables the DCGM health check system for the specified systems;

* HealthGet - Retrieves the current state of the DCGM health check system;

* HealthCheck - Checks the configured watches for any errors, failures, or warnings.

### CODE REVIEW

**Review by glowkey — APPROVED** [review]
LGTM

File: pkg/dcgm/health.go [hunk 1]
```diff
@@ -23,14 +40,84 @@ type DeviceHealth struct {
...
+	System     HealthSystem
+	Health     HealthResult
+	Error      IncidentError
+	EntityInfo GroupEntityPair
+}
+
+type HealthResponse struct {
+	OverallHealth HealthResult
+	Incidents     []Incident
+}
```
Comment (rohit-arora-dev) [inline pkg/dcgm/health.go:78]: I like how you are capturing the right details, however some details like `Code` and `EntityInfo` are not used today. Is it done to re-use the HealthResponse struct to provide more details in future?

File: pkg/dcgm/health.go [hunk 2]
```diff
@@ -23,14 +40,84 @@ type DeviceHealth struct {
...
+	var systems C.dcgmHealthSystems_t
+
+	result := C.dcgmHealthGet(handle.handle, groupId.handle, (*C.dcgmHealthSystems_t)(unsafe.Pointer(&systems)))
+	if err := errorString(result); err != nil {
+		return HealthSystem(0), err
+	}
+	return HealthSystem(systems), nil
+}
+
+type IncidentError struct {
```
Comment (rohit-arora-dev) [inline pkg/dcgm/health.go:63]: I would appreciate if you include a `Error() string` implementation as suggested by GoLang error interface:
```
type error interface {
  Error() string
}

```

**Review by rohit-arora-dev — COMMENTED** [review]
Overall looking good, just a few minor suggestions.

File: pkg/dcgm/health.go [hunk 3]
```diff
@@ -23,14 +40,84 @@ type DeviceHealth struct {
...
+	var healthResults C.dcgmHealthResponse_v4
+	healthResults.version = makeVersion4(unsafe.Sizeof(healthResults))
+
+	result := C.dcgmHealthCheck(handle.handle, groupId.handle, (*C.dcgmHealthResponse_t)(unsafe.Pointer(&healthResults)))
+
+	if err := errorString(result); err != nil {
+		return HealthResponse{}, &DcgmError{msg: C.GoString(C.errorString(result)), Code: result}
+	}
+
+	response := HealthResponse{
```
Comment (nvvfedorov) [inline pkg/dcgm/health.go:94]: @rohit-arora-dev I removed unused structure. Thank you for your finding.

File: pkg/dcgm/health.go [hunk 4]
```diff
@@ -23,14 +40,84 @@ type DeviceHealth struct {
...
+	result := C.dcgmHealthGet(handle.handle, groupId.handle, (*C.dcgmHealthSystems_t)(unsafe.Pointer(&systems)))
+	if err := errorString(result); err != nil {
+		return HealthSystem(0), err
+	}
+	return HealthSystem(systems), nil
+}
+
+type DiagErrorDetail struct {
+	Message string
+	Code    HealthCheckErrorCode
```
Comment (nvvfedorov) [inline pkg/dcgm/health.go:65]: @rohit-arora-dev , The IncidentError was renamed to DiagErrorDetail that is match with the name of the DCGM original structure.

File: pkg/dcgm/health.go (hunk 1 above)
Comment (nvvfedorov) [reply pkg/dcgm/health.go]: @rohit-arora-dev , The goal to expose the full response to go-dcgm package consumers, such as the dcgm-exporter.

**Review by glowkey — APPROVED** [review]

**Pull request merged**